cupy/cuda/cub.pyx

```diff
@@ -60,13 +60,25 @@ def reduce_sum(core.ndarray x, out=None):
     ws = core.ndarray(ws_size, numpy.int8)
     ws_ptr = <void *>ws.data.ptr
     cub_reduce_sum(x_ptr, y_ptr, x.size, ws_ptr, ws_size, dtype_id)
+    if keepdims:
+        y = y.reshape((1,))
     if out is not None:
+        if y.ndim != out.ndim:
+            raise ValueError(
+                "output parameter for reduction operation sum has the wrong "
+                "number of dimensions")
```

Comment:
If we are gonna raise, can we do it before calling `cub_reduce_*()`? Let's not waste time on computation if the result isn't valid.

Reply:
Sure. At the moment the check is simple since we don't have axis support yet. It is basically just checking that `out.ndim == keepdims`